[RFC] Bugfixes

tarruda opened the pull request:
Fix #3188 #2322 #3197

tarruda committed:
tui/remote_ui: Fix some regressions

- Explicitly set the SignalWatcher event queue. Without this, the watcher will
  publish events to the fast queue, resulting in resize bugs for certain
  terminals(#2322).
- Set `async = false` to the `remote_ui_attach` handler(It was a deferred
  before, this is the new equivalent)

tarruda committed:
main: Initialize event loop before command_line_scan

The call to `event_init()` was too late. `command_line_scan()` in `main()` could
already need the loop initialized. Ref https://github.com/neovim/neovim/issues/3045#issuecomment-123405833.

A consequence of this change is that it was necessary to move the
`channel_from_stdio()` call to `command_line_scan()` when embedded_mode is set.

oni-link commented on src/nvim/tui/tui.c:
```diff
@@ -140,6 +140,7 @@ UI *tui_start(void)
 
   // listen for SIGWINCH
   signal_watcher_init(&loop, &data->winch_handle, ui);
+  data->winch_handle.events = queue_new_child(loop.events);
```
That queue is never freed (when calling tui_stop()).

tarruda replied:
:+1:

tarruda committed:
eval: Protect job callbacks from being redefined

ref: #3188

justinmk commented:
Can these be tested? I may have repro for the job callback issue, when I get back to a computer later.

tarruda committed:
terminal: Only wipe buffer terminal_close wasn't called

After @250aca4f8938 it is possible that terminal_close will be called without
invoking the close_cb(which normally destroys the terminal structure). If this
happens, the terminal buffer will already be deleted so there's no need to call
`bwipeout!`.

tarruda commented:
> Can these be tested? I may have repro for the job callback issue, when I get back to a computer later.

I've added a test for the job callback change. As for the others:

- The signal watcher change only happens in certain terminals and is not applicable to the lua UI, so can't automate the test
- I couldn't find a way to deterministically reproduce the `remote_ui_attach` issue
- The `event_init` change is possible to test, but might require some adapting to our test infrastructure(we need a way to pass specific command-line flags to individual tests)
- The terminal change I was only able to reproduce with FZF

justinmk commented on test/functional/job/job_spec.lua:
```diff
@@ -188,6 +189,41 @@ describe('jobs', function()
+      \ 'on_stderr': function('g:JobHandler'),
+      \ 'on_exit': function('g:JobHandler')
+      \ }
+      let job = jobstart('cat -', g:callbacks)
```
on windows, `pause` might be a good substitue for `cat -`.

tarruda replied:
When we run tests on windows, it might be a good a idea to use msys to provide a minimal set of unix command line tools
